Mobile

FedericoDiTosto opened the pull request:
I created a new button that allows you to view ,through a new preview canvas, the next block to be inserted, change its shape, color and rotation to improve usability from mobile. 
Made the options bar responsive for mobile viewing.

FedericoDiTosto committed:
sync upstream

FedericoDiTosto committed:
Merge remote-tracking branch 'upstream/working' into working

FedericoDiTosto committed:
sync

FedericoDiTosto committed:
MobileMode & Responsive

AlasdairMott commented:
that's great,  makes it so much easier on mobile!

The menu animation is a little jerky, I reckon leave it out for now. It's also on top of the help modal, as is the mobile icon. 

https://user-images.githubusercontent.com/25563790/200143874-7b9e529c-767b-4b6e-9e4b-45947445b032.mp4

I was thinking it could look something like this: 

<img width="600px" alt="image" src="https://user-images.githubusercontent.com/25563790/200144104-f8955ced-8f31-4665-8f80-291e0781de04.png">

On mobile the icons spill over. We can have it scale or maybe share should sit in the lower right with help and mute?

<img width="414" alt="image" src="https://user-images.githubusercontent.com/25563790/200144193-46323267-284e-4333-b042-61ea36402c11.png">

AlasdairMott commented on src/3d/Viewport3d.tsx:
```diff
@@ -152,14 +154,17 @@ const Viewport3d = () => {
-            <HelpAndSoundButtons onShowHelp={() => setShowHelp(!showHelp)}/>
-            {showHelp && <HelpModal onClose={() => setShowHelp(false)}/>}
+            <HelpAndSoundButtons onShowHelp={() => setShowHelp(!showHelp)} />
+            {showHelp && <HelpModal onClose={() => setShowHelp(false)} />}
```
move 163 and 164 after the mobile buttons to that the help modal displays on top :)

AlasdairMott commented on src/UI/MobileSelector.module.css:
```diff
@@ -0,0 +1,43 @@
+    border-radius: 12px;
+}
+
+.previwButtons {
```
typo here

AlasdairMott commented on src/UI/MobileMode.tsx:
```diff
@@ -0,0 +1,51 @@
+            <div className={styles.canvasContainer}>
+                <Canvas shadows={true} camera={{ position: [1, 1, 1] }}>
+                    <color attach="background" args={[1, 1, 1]} />
+                    <OrbitControls />
```
not sure if the preview will need orbit controls, adds some complexity if the user pans or zooms out far

AlasdairMott commented on src/UI/MobileMode.tsx:
```diff
@@ -0,0 +1,51 @@
+            <div className={`${styles.previwButtons} glass`}>
+                <ToolbarCommandButton onClick={() => { console.log(props.cam); dispatch(modelActions.nextBlock()) }} title={"next-shape"}><AddHexagon {...ICONPROPS} />
+                </ToolbarCommandButton>
+                <ToolbarCommandButton onClick={() => { dispatch(modelActions.rotateBlock('right')) }} title={"change-color"}><CropRotateTr {...ICONPROPS} />
```
Its not ideal but I reckon iconoir`refresh-double` would be clearer here

FedericoDiTosto commented:
Great idea, the menu is much cleaner this way. I will fix it as soon as I can. 
I tried to make the mobile menu canvas rotate every time you move with the OrbitControl the main scene but without success(so that the user knows exactly at what angle the new block will be positioned). Do you know how to make two canvas rotate with only one OrbitControl?

FedericoDiTosto committed:
fixing menu layout

FedericoDiTosto commented:
I've fixed the layout, let me know if I've forgotten anything or you have new ideas. For the top toolbar I fixed (only for some devices) by decreasing the side margin obviously it is a temporary solution.

AlasdairMott commented:
Wow that looks so good! Looks great on my mobile.

> Do you know how to make two canvas rotate with only one OrbitControl?

I tried doing something like this to keep the perspective and ortho cameras in sync, but in the end I had to only sync when swapping over. 

A workaround might be an xyz gizmo in the viewport + preview.
